cachi2/core/models/output.py

```diff
@@ -0,0 +1,113 @@
+from pathlib import Path
+from typing import Any, Callable, Iterable, Literal, Optional, TypeVar
+
+import pydantic
+
+T = TypeVar("T")
+
+
+def unique(items: Iterable[T], by: Callable[[T], Any], dedupe: bool = True) -> list[T]:
+    """Make sure input items are unique by the specified key.
+
+    The 'by' function must return a hashable key (the uniqueness key).
+
+    If item A and item B have the same key, then
+        if dedupe is true (the default) and A == B, B is discarded
+        if dedupe is false or A != B, raise an error
+    """
+    by_key = {}
+    for item in items:
+        key = by(item)
+        if key not in by_key:
+            by_key[key] = item
+        elif not dedupe or by_key[key] != item:
+            raise ValueError(f"conflict by {key}: {by_key[key]} X {item}")
+    return list(by_key.values())
+
+
+def unique_sorted(items: Iterable[T], by: Callable[[T], Any], dedupe: bool = True) -> list[T]:
+    """Make sure input items are unique and sort them.
+
+    Same as 'unique()' but the key returned from the 'by' function must support ordering.
+    """
+    unique_items = unique(items, by, dedupe)
+    unique_items.sort(key=by)
+    return unique_items
+
+
+# Supported package types (a superset of the supported package *manager* types)
+PackageType = Literal["gomod", "go-package"]
+
+
+class Dependency(pydantic.BaseModel):
+    """Metadata about a resolved dependency."""
+
+    type: PackageType
+    name: str
+    version: Optional[str]  # go-package stdlib dependencies are allowed not to have versions
+
+    @pydantic.validator("version")
+    def must_have_version_unless_gopackage(
+        cls, version: Optional[str], values: dict
+    ) -> Optional[str]:
+        """Check that the dependency has a version or is 'go-package'."""
+        ptype = values.get("type")
+        if ptype is not None and (version is None and ptype != "go-package"):
+            raise TypeError(f"{values['type']} dependencies must have a version")
+        return version
+
+
+class Package(pydantic.BaseModel):
+    """Metadata about a resolved package and its dependencies."""
+
+    type: PackageType
+    path: Path  # relative from source directory
+    name: str
+    version: str
+    dependencies: list[Dependency]
+
+    @pydantic.validator("path")
+    def path_is_relative(cls, path: Path) -> Path:
+        """Check that the package path is relative."""
+        if path.is_absolute():
+            raise ValueError(f"package path must be relative: {path}")
+        return path
+
+    @pydantic.validator("dependencies")
+    def unique_deps(cls, dependencies: list[Dependency]) -> list[Dependency]:
+        """Sort and de-duplicate dependencies."""
+        return unique_sorted(dependencies, by=lambda dep: (dep.type, dep.name, dep.version))
+
+
+class EnvironmentVariable(pydantic.BaseModel):
+    """An environment variable."""
+
+    name: str
+    value: str
+    kind: Literal["literal", "path"]
```

Comment:
How are we gonna do environment variables? Previously, OSBS had to take `path` env. vars and make their values absolute.

Do we want to do that right away and get rid of the `kind`? Do we want a subcommand that generates fully resolved env. vars in the specified format (json, .env, ...)? Do we want to decide later?

Reply:
The fully resolved paths would be based on the `output_dir`? I'm trying to imagine how this would be used in a Tekton pipeline, and I think the build step can change the `output_dir` path when it mounts it.

On the other hand, maybe we can see this from the perspective of: you said the output should be "here", so we're delivering env variables considering that output is "here", if you change where output is, you should handle the env variables. Sounds more simple, and can be enough for an initial approach.

Reply:
Yeah, same issue with mismatched paths that I mentioned in the pip design. I'll hack together what the subcommand could look like and we can discuss it later